rep/publisher.go

```diff
@@ -3,33 +3,54 @@ package rep
 import "sync"
 
 type RetrievalEventPublisher struct {
-	mu          sync.Mutex
-	subscribers map[RetrievalSubscriber]struct{}
+	subscribersLk  sync.RWMutex
+	subscriberList []RetrievalSubscriber
 }
 
-func NewEventPublisher() (*RetrievalEventPublisher, error) {
-	return &RetrievalEventPublisher{
-		subscribers: map[RetrievalSubscriber]struct{}{},
-	}, nil
+func NewEventPublisher() *RetrievalEventPublisher {
+	return &RetrievalEventPublisher{subscriberList: make([]RetrievalSubscriber, 0)}
 }
 
 func (ep *RetrievalEventPublisher) SubscribeToRetrievalEvents(subscriber RetrievalSubscriber) {
-	// mutex lock writes
-	ep.mu.Lock()
-	defer ep.mu.Unlock()
-	ep.subscribers[subscriber] = struct{}{}
+	// Lock writes on the subscribers list
+	ep.subscribersLk.Lock()
+	defer ep.subscribersLk.Unlock()
+
+	for _, subInList := range ep.subscriberList {
+		// Don't add the subscriber again if they're already in the list
+		if subInList == subscriber {
+			return
+		}
+	}
+
+	// Add the subscriber to the list
+	ep.subscriberList = append(ep.subscriberList, subscriber)
 }
 
 func (ep *RetrievalEventPublisher) UnsubscribeFromRetrievalEvents(subscriber RetrievalSubscriber) {
-	// mutex lock writes
-	ep.mu.Lock()
-	defer ep.mu.Unlock()
-	delete(ep.subscribers, subscriber)
+	// Lock writes on the subscribers list
+	ep.subscribersLk.Lock()
+	defer ep.subscribersLk.Unlock()
+
+	curLen := len(ep.subscriberList)
+	for i, subInList := range ep.subscriberList {
+		// Remove the subscriber if they're in the list
+		if subInList == subscriber {
```

Comment:
So I hate to say this, but this doesn't fundamentally solve the problem with using the subscriber as a map key.

The basic problem here is the == operator will not work with all types. only those that are comparable -- https://go.dev/ref/spec#Comparison_operators

You're probably working from a javascript viewpoint, where everything is an object and you can always call == or === on them. In go, some types will panic when you do this. Unfortunately it's only at run time you realize this when you get the panic -- there's no compile error. (I'd argue this is a pretty terrible design flaw in the language, but not much to be done on that).

Again, you're likely going to get an error when you try to get this going in run time. I'll provide a test you can work off.

Reply:
Oh, duh. I get it. You're saying the Subscriber itself is not comparable. You're right then I didn't really change anything because I end up comparing the subscribers anyway. I had it in my head when writing this that the problem was the keyed map because a Subscriber is not comparable, then blindly trying to compare it, haha.